### DISCUSSION TITLE
Set window position?

### DISCUSSION THREAD

**Discussion by emendelson** [dosemu2/dosemu2#1892] [Q&A]
Is there a way to specify the location on screen where the doseum2 window opens? On my system, it always opens on the lower left of the screen. Is there a way to specify that the upper-left corner of the window should be at location x,y (for example, 75, 100)?

Probably the answer to this is something everyone knows, but I haven't found it. Thanks for any help.

**Comment by stsp** [comment]
man devilspie2
Well, man is quite terse there but
you may google out the examples
for devilspie2.

**Comment by emendelson** [comment]
Thank you - I've been experimenting with the the Windows Subsystem for Linux, which doesn't (by default) run gnome, so I can't use devilspie2, but I see that this may be possible now with some third-party software. Will experiment.

**Comment by stsp** [comment]
You mean wsl2 on windows-10?
How good is dosemu2 there?
I once got it to work on wsl1 but it wasnt good.

**Comment by emendelson** [comment]
Yes wsl2 running Ubuntu on Windows 11. Works extremely well - amazingly fast, and faster than DOSBox running natively. The only problems are annoying messages when I start it:

```
ERROR: KVM: error opening /dev/kvm: Permission denied
ERROR: userspace pings are disabled.
        To enable them, do as root:
        echo 0 65535 > /proc/sys/net/ipv4/ping_group_range
```

I thought I figured out how to fix the first message, but it came back after restarting. If I enter `ls /dev/kvm`, it returns this:

`[1]+  Done                  dosemu`

which I don't understand at all.

About the second error, I logged into Ubuntu as root and performed the suggested command, but the error message came back today. If you can suggest a way around them, I'll be grateful.

**Reply by stsp** [reply]
> Works extremely well - amazingly fast, and faster than DOSBox running natively.

Even without kvm?
Very interesting. :)
Does sound and midi work?

> [1]+  Done                  dosemu

I think you should ignore that and
just re-type `ls -l /dev/kvm`.

I silenced the error about userspace pings.

**Comment by emendelson** [comment]
I reinstalled qemu-kvm and ran `sudo adduser <username> kvm` and the kvm error message went away... A case of user-stupidity. 

And thank you for silencing the userspace pings message....

I've got sound working in WSL (listening to YouTube in Chrome right now), but when I try to play a MIDI clip in WordPerfect 6.2, I get a "hardware error" message from WP. The same setup works in non-WSL Linux. I haven't modified any of the default sound settings in dosemu in either setup. 

Do you have a simple sound-producing program that I can try instead of difficult WordPerfect?

**Reply by stsp** [reply]
If you have `install-freedos` package
installed and run `insfdusr` then you
would have `opencp` - a sound player.

**Reply by stsp** [reply]
> I reinstalled qemu-kvm and ran

What distro is it?
I thought installing qemu-kvm to get
the right permissions, is not needed
for 5 years now.

**Reply by stsp** [reply]
And did it make any faster now?

**Reply by stsp** [reply]
Also consider producing the -D9+S
log for that "hardware error". Maybe
its something simple.

**Comment by emendelson** [comment]
I'll get the -D9+S log later today or tomorrow (and try to guess whether it's faster), but, meanwhile to answer your other question: the distro is Ubuntu 22.04.1 LTS running in WSL2 under Windows 11.

**Reply by stsp** [reply]
So the need of qemu-kvm for kvm
perms in 22.04 is a bit unexpected.
Maybe someone else has an idea
how can that be. IIRC all you need
is a recent enough udev.
So what udev version is there?

**Comment by emendelson** [comment]
When I enter `udevadm --version` it returns `249`. Does that tell you anything? This is all new territory to me...

**Reply by stsp** [reply]
Yes, 240 was enough in the past...

**Comment by emendelson** [comment]
And I hope this is the right log. I started dosemu like this:

`dosemu -D -D9+S`

and then ran WP, opened a file with a MIDI clip, and tried to play it:

[boot.log](https://github.com/dosemu2/dosemu2/files/10735526/boot.log)

If I got this wrong, let me know exactly what I should be doing.

**Reply by stsp** [reply]
Remove one extra -D

**Comment by emendelson** [comment]
Ah - I should have understood:

[boot.log](https://github.com/dosemu2/dosemu2/files/10735854/boot.log)

**Reply by stsp** [reply]
I think you have sound disabled in
the config. Please show -D9+cS log.

**Comment by emendelson** [comment]
Many apologies for wasting your time. I simply forgot that I had disabled sound, which I think I did because of error messages (see below)

I re-enabled sounddos and tried again, and this time got a two-minute long "Please wait" message from WordPerfect before it finally played the midi file. It also produced these errors on launch (and the /dev/kvm error is back):

```
ERROR: KVM: error opening /dev/kvm: Permission denied
ERROR: alsa_midi:virtual (ALSA err 2): open /dev/snd/seq failed: No such file or directory
snd-virmidi module not loaded or device "virtual" not configured
see "amidi -l" for the list of midi devices
```
And here's the log:

[boot.log](https://github.com/dosemu2/dosemu2/files/10737680/boot.log)

Is there any hope of silencing those error messages? And apologies again for wasting your time earlier. I should have known better.

**Reply by stsp** [reply]
> I re-enabled sounddos and tried again, and this time got a two-minute long "Please wait"

Maybe you forgot to apply the patch
to wordperfect that removes the midi
playing delay? Or maybe that's because
you had dosemu -D9+S logging enabled.
Or maybe because of no kvm.
All these possibilities should be verified
separately.

> Is there any hope of silencing those error messages?

As for kvm error you still haven't showed
`ls -l /dev/kvm` so nothing can be said.
Maybe you need to add yourself to the
"kvm" group.

As for midi, you can update dosemu and
add `$_snd_plugin_params = "alsa_virmidi:enabled=0"`
to disable that error.

**Comment by emendelson** [comment]
Thank you for this, and apologies for not answering earlier questions.

Output of `ls -l /dev/kvm`:

`crw------- 1 root root 10, 232 Feb 14 18:10 /dev/kvm`

I seem to be a member of the kvm group already:

```
edward@TC920-Win11:~$ sudo adduser edward kvm
The user `edward' is already a member of `kvm'.
```

I updated dosemu and added that line to .dosemu/.dosemurc, and still get

```
ERROR: KVM: error opening /dev/kvm: Permission denied
ERROR: alsa_midi:virtual (ALSA err 2): open /dev/snd/seq failed: No such file or directory
snd-virmidi module not loaded or device "virtual" not configured
see "amidi -l" for the list of midi devices
```

About the patch to WordPerfect. I thought I knew everything about WPDOS, but this is new to me. I'm using 6.2. Which patch are you referring to?

Also, if it helps, here is my .dosemurc (renamed dosemu.txt so that I can upload it):

[dosemurc.txt](https://github.com/dosemu2/dosemu2/files/10745469/dosemurc.txt)

**Reply by stsp** [reply]
> crw------- 1 root root 10, 232 Feb 14 18:10 /dev/kvm

These are the wrong pers, so
there is some problem with udev.
You should report that to ubuntu
(for example) because the kvm
perms are not dosemu2-specific.

> I updated dosemu and added that line to .dosemu/.dosemurc, and still get

Maybe it didn't update in PPA yet,
and as you do not post the version,
there is nothing to suggest but to
wait and try update tomorrow again.

> Which patch are you referring to?

The one from your howto.

**Reply by emendelson** [reply]
The patch: someday I really should look at my own site... The sound you hear is me slapping my forehead. I've copied the patched file into my WP directory, and sound now starts instantly. Apologies for user-stupidity on a massive scale.

Version number:

```
dosemu2-2.0pre9
Revision: 7076
```

Will report the kvm issue to Ubuntu. Thank you.

EDIT: The kvm issue is a known problem because (it says somewhere) wsl doesn't have udev. The solution (tested by me) is this:

```
# /etc/wsl.conf
[boot]
command = /bin/bash -c 'chown root:kvm /dev/kvm && chmod 660 /dev/kvm'
```

Edit the file as sudo to add those lines, and restart wsl. Problem solved.

**Reply by stsp** [reply]
> EDIT: The kvm issue is a known problem because (it says somewhere) wsl doesn't have udev.

I don't understand what does this mean,
but I've found that silly note too...
I can only guess that because wsl2 uses
some thin hypervisor, perhaps it messes
up device IDs sometimes, and so udev
can't work... still weird. But that's MS. :)

> Version number:

Bad number, see `dpkg-query -s dosemu2`
for a better one.
Or better just let me know if the problem
persists tomorrow.

**Reply by emendelson** [reply]
OK, Meanwhile, I've rewritten my page to recommend (as you do) creating an empty .dosemurc file and adding lines to it. And, while we're here, the version number I got today seems to say that it's still yesterday's build, if that's what `20230214` means:

`2.0~pre9-7904-acb857e30+202302142016~ubuntu22.04.1` 

Will report back tomorrow.

EDIT: And it's still frustrating that dosemu2 always opens with 3/4 of its window in my left-hand monitor and 1/4 in my right-hand monitor. I have it set to open with 43 lines, and the bottom third of the window is invisible, below the monitors. Probably there's no way around this with WSL, but it would be nice if dosemu2 could open near the upper left of the monitor (or the two monitors).

EDIT2: New version arrived, and the alsa error messages are gone. Thank you.

**Reply by stsp** [reply]
I've found this:
https://ubuntu.com/tutorials/install-ubuntu-on-wsl2-on-windows-11-with-gui-support#4-configure-ubuntu
It says how to enable systemd.
I suppose this would also enable udev?

**Reply by emendelson** [reply]
I enabled systemd and restarted, but nothing changed. I already have dosemu opening with no error messages, presumably because of what I described in earlier posts here, including one that said: 

I reinstalled qemu-kvm and ran `sudo adduser <username> kvm` and the kvm error message went away.

So right now, the only annoyance is that dosemu always opens with its window partly in my left-hand monitor and partly in the right-hand monitor. Since wsl doesn't use a graphical desktop, I can't figure out how to change this. Again, it would be good to have a setting in .dosemurc that would let me set the x,y position of the upper right corner of the window, but maybe that's simply not possible.
